src/kOS.Safe/Encapsulation/ScalarValue.cs

```diff
@@ -0,0 +1,456 @@
+using kOS.Safe.Exceptions;
+
+namespace kOS.Safe.Encapsulation
+{
+    abstract public class ScalarValue : Structure, IConvertible
+    {
+        protected object internalValue;
+
+        abstract public bool IsInt { get; }
+
+        abstract public bool IsDouble { get; }
+
+        public bool IsValid
+        {
+            get
+            {
+                if (IsInt) return true;
+                if (IsDouble)
+                {
+                    double d = GetDoubleValue();
+                    if (double.IsInfinity(d) || double.IsNaN(d)) return false;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public object Value { get { return internalValue; } }
+
+        protected ScalarValue()
+        {
+            InitializeSuffixes();
+        }
+
+        public void InitializeSuffixes()
+        {
+            AddSuffix("ISINT", new Suffix<bool>(() => IsInt));
+            AddSuffix("ISDOUBLE", new Suffix<bool>(() => IsDouble));
+            AddSuffix("ISVALID", new Suffix<bool>(() => IsValid));
+        }
+
+        public static ScalarValue Create(object value)
+        {
+            if (value is float)
+                value = Convert.ToDouble(value);
+            if (value is double)
+            {
+                bool inBounds = Int32.MinValue < (double)value && (double)value < Int32.MaxValue;
+                if (inBounds && !double.IsNaN((double)value))
+                {
+                    // Convert the double to an int, and check and see if they are still equal.
+                    // if so, treat the double as if it was an int.
+                    int intPart = Convert.ToInt32(value);
+                    if ((double)value == intPart)
+                    {
+                        return new ScalarIntValue(intPart);
+                    }
+                }
+                return new ScalarDoubleValue((double)value);
+            }
+            else if (value is int)
+            {
+                return new ScalarIntValue((int)value);
+            }
+            else if (value is ScalarValue)
+            {
+                return ScalarValue.Create(((ScalarValue)value).Value);
+            }
+            else
+            {
+                throw new kOS.Safe.Exceptions.KOSException(string.Format("Failed to set scalar value.  Passed type {0}, expected Double or Int", value.GetType().Name));
+            }
+        }
+
+        public int GetIntValue()
+        {
+            return Convert.ToInt32(Value);
+        }
+
+        public double GetDoubleValue()
+        {
+            return Convert.ToDouble(Value);
+        }
+
+        public override string ToString()
+        {
+            if (IsInt) return GetIntValue().ToString();
+            else if (IsDouble) return GetDoubleValue().ToString();
+            return "NaN";
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null) return false;
```

Comment:
The null check prior to casting is needed for null safety.  I should remove the `is` however, since the `as` and `!= null` are equivalent, which means we're casting twice instead of once.